File: Java/373.java
__________________________________________________________________________________________________
sample 1 ms submission
class Solution {
    public List<int[]> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        List<int[]> res = new ArrayList<>();
        if (nums1.length == 0 || nums2.length == 0 || k <= 0) return res;
        if (nums1.length <= k/nums2.length){
            for(int n1: nums1){
                for(int n2: nums2){
                    res.add(new int[]{n1,n2});
                }
            }
            return res;
        }
        int lo = nums1[0] + nums2[0], hi = nums1[nums1.length - 1] + nums2[nums2.length - 1];
        
        while(lo < hi){
            int mid = lo + ((hi - lo) >> 1);
            int count = 0;
            for(int i = 0; i < nums1.length && count < k; i++) {
                count += binarySearch(nums2, 0, nums2.length, mid - nums1[i]);
            }
            if(count >= k) hi = mid;
            else lo = mid + 1;
        }
        
        for (int i = 0; i < nums1.length; i++){
            for (int j = 0; j < nums2.length && nums1[i] + nums2[j] < lo; j++) {
                res.add(new int[]{nums1[i], nums2[j]});
            }
        }
        for (int i = 0; i < nums1.length; i++){
            for (int j = 0; j < nums2.length && nums1[i] + nums2[j] <= lo; j++) {
                if (nums1[i] + nums2[j] == lo) {
                    res.add(new int[]{nums1[i], nums2[j]});
                }
                if (res.size() == k) {
                    return res;
                }
            }
        }
        return res;
    }
    
    int binarySearch(int[] nums, int left, int right, int target) {
        while(left < right){
            int mid = left + (right - left) / 2;
            if(nums[mid] > target)  {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return right;
    }
}
__________________________________________________________________________________________________
sample 35708 kb submission
public class Solution {
    public List<int[]> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        PriorityQueue<int[]> que = new PriorityQueue<>((a,b)->a[0]+a[1]-b[0]-b[1]);
        List<int[]> res = new ArrayList<>();
        // Handle edge case
        if(nums1.length==0 || nums2.length==0 || k==0) return res;
        
        // k because we need top k?
        for(int i=0; i<nums1.length && i<k; i++) que.offer(new int[]{nums1[i], nums2[0], 0});
        
        while(k > 0 && !que.isEmpty()){
            int[] cur = que.poll();
            res.add(new int[]{cur[0], cur[1]});
            k--;
            if(cur[2] == nums2.length-1) continue;
            que.offer(new int[]{cur[0],nums2[cur[2]+1], cur[2]+1});
        }
        return res;
    }
}
__________________________________________________________________________________________________

File: Python3/373.py
__________________________________________________________________________________________________
sample 40 ms submission
import heapq
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        if not nums1 or not nums2:
            return []
        n,m=len(nums1),len(nums2)
        mappings=[0]*n
        que=[[nums1[i]+nums2[0], i] for i in range(n)]
        res=[]
        while len(res)<k and que:
            val,idx=heapq.heappop(que)
            res.append([nums1[idx],nums2[mappings[idx]]])
            mappings[idx]+=1
            if mappings[idx]<m:
                val2=nums1[idx]+nums2[mappings[idx]]
                heapq.heappush(que,[val2,idx])
        return res
__________________________________________________________________________________________________
sample 13304 kb submission
import heapq
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        if not nums1 or not nums2:
            return []
        
        def push(i, j):
            heapq.heappush(heap, (nums1[i] + nums2[j], i, j))
        
        def pop():
            _, i, j = heapq.heappop(heap)
            return i, j
        
        res = []
        heap = [(nums1[0] + nums2[0], 0, 0)]
        while heap and len(res) < k:
            i, j = pop()
            res.append((nums1[i], nums2[j]))
            if j == 0 and i + 1 < len(nums1):
                push(i + 1, j)
            if j + 1 < len(nums2):
                push(i, j + 1)
        
        return res
__________________________________________________________________________________________________
